### PR TITLE
feat(rollout/sglang_diffusion): support ti2i on FLUX.2-Klein

### PR DESCRIPTION
**Pull request by celve**
> **Rebased onto the Sample/Part agentic migration (2026-07-28).** Main has adopted
> the `Sample`/`Part` model (`RolloutReq`/`RolloutResp` retired). This PR is
> re-implemented on top of it and force-pushed in place. The practical consequence:
> the earlier `task_config`/`stage_config` compatibility layer added in review is
> **gone** — there is no `stage_config` to migrate anymore. The ti2i branch is now
> driven by `Sample.has_image_input()` and the edit dataset, so no task flag or
> cross-PR (#255) coupling remains. The single commit rebases cleanly on current
> `main` (base was an exact match at force-push time).

## Summary

Adds text+image→image (ti2i) support to the `sglang_diffusion` rollout engine for
FLUX.2-Klein, so image-editing GRPO can run out-of-process instead of only through
`TrainsideRolloutEngine`.

Klein serves both modalities off one checkpoint, so this extends the existing
`flux2_klein` adapter rather than registering a second edit key. Every layer already
branches on *image presence* rather than a mode flag — upstream SGLang's pipeline
config is `ModelTaskType.TI2I` unconditionally and its `ImageVAEEncodingStage` no-ops
without a condition image; the trainside `Flux2KleinPipeline` branches on
`req.primitives['image']`; and `Flux2KleinConditions` already declares
`image_latent` / `image_latent_ids` optional and round-trips both. **The trainside
half needs no changes.**

- `Flux2KleinAdapter.build_prompts(sample)` returns the inherited t2i payload verbatim
  when `not sample.has_image_input()`; otherwise it pulls the source PIL from
  `sample.vision_conditioning()` and ships it via the `condition_image` sampling kwarg.
  The `patch_sampling_io` plumbing built for Edit-Plus is not gated on model family and
  already handles per-prompt indexing and preserving the driver-pinned output size.
  Images collapse in lockstep with the prompt collapse (`first_per_group`), so image
  *i* stays paired with prompt *i* under interleaved `group_ids`.
- `build_condition` emits the packed condition tokens (`image_latent`) plus their 4-axis
  RoPE ids (`condition_image_latent_ids`) as two `ImageLatentCondition` entries. No
  unpack is needed — FLUX.2 returns the packed `[B, N, 128]` token form the trainside
  slot already wants. The ids are **captured, not recomputed**: `N` alone does not give
  the `h_pat × w_pat` factorization, and FLUX.2 never populates `vae_image_sizes` (it
  does not override the base `preprocess_vae_image` no-op), so deriving the grid
  driver-side would duplicate upstream's condition-image resize math and desync replay
  whenever it changes.
- `patch_conditions` captures `condition_image_latent_ids` alongside `image_latent`,
  riding the existing list-based merge/slice path with no special case (it is a uniform
  `[B, N, 4]` per-sample tensor, unlike the `image_latent_sizes` pair Edit-Plus needs).
- `build_segment` is inherited unchanged. SGLang concatenates condition tokens into a
  separate `latent_model_input` and never writes them back to `ctx.latents`, so the
  recorded trajectory stays noise-only and the existing Klein packed-token unpack is
  correct for both t2i and ti2i.
- Recipes: `flux2_klein_editreward_sglang.yaml` (base-9B) and
  `flux2_klein_4b_editreward_sglang.yaml` (base-4B) — the SGLang siblings of the proven
  trainside `flux2_klein_editreward.yaml` / `flux2_klein_4b_editreward.yaml`.

## Related Issue

N/A

### A. Migrated engine (this PR's code) — validated on GPU now

**GPU round-trip through `SGLangDiffusionRolloutEngine`** — 1×H20, FLUX.2-Klein-4B,
8 steps @ 512², `guidance_scale=1.0`, `eta=0.7`, `sde_indices=[0,1,2]`, K=2,
`use_lora=False`. The harness builds `Sample`s (`Part.input` + `input_child({'image': …})`,
`Sample.request`, `sample.fork(...)`) and drives them through the engine directly (no
trainer), then reads `result.frontier_gen_part(DiffusionSamplingParams).conditions`.
**24/24 checks passed.**

- **T1 — ti2i, 1 prompt × K=2** → `image_latent (2, 1024, 128)` +
  `image_latent_ids (2, 1024, 4)`; the ids are **exactly**
  `cartesian_prod(10, arange(32), arange(32), arange(1))` (`torch.equal`), confirming
  SGLang's `_prepare_image_ids(scale=10)` matches trainside `REFERENCE_TIME_SCALE = 10`;
  the two K-samples share one source latent.
- **T2 — ti2i, 2 prompts × K=2** → `image_latent (4, 1024, 128)` + `(4, 1024, 4)`;
  samples within a group share the source latent, different groups do not — per-prompt
  `condition_image` indexing under the grouped merge/slice path is correct.
- **T3 — t2i non-regression** (no image) → conditions are `['text']` only; the image
  branch is a true no-op, and `build_condition` emits neither `image_latent` nor
  `image_latent_ids`.

**Recipe + static** (both recipes, `.venv-sglang`, migrated `main`):
- `python -m unirl.train_diffusion --config-name=diffusion/flux2_klein/flux2_klein_editreward_sglang --cfg job` → composes; `model_family=flux2_klein`, `populate_conditions=true`, `init_same_noise=false`, `base-9B`, `batch_size=48`.
- same for `…/flux2_klein_4b_editreward_sglang --cfg job` → `base-4B`, `batch_size=64`.
- `python scripts/check_recipe_targets.py` → **2193 `_target_` paths resolve** (both new recipes included).
- `ruff check` + `ruff format --check` on `adapters/flux.py` and `_patches/patch_conditions.py` → clean.

### B. Engine-equivalence A/Bs — pre-migration implementation of this same path

> These two multi-hour A/Bs were run on the **pre-migration** implementation of this
> exact ti2i path (before `main` adopted Sample/Part). The migration is a mechanical
> port of the type plumbing (`RolloutReq` → `Sample`); the ti2i logic — ship
> `condition_image` → capture `image_latent` + `condition_image_latent_ids` → replay —
> is unchanged, and **§A confirms the migrated code emits the identical conditioning
> tensors**. I did not re-run the A/Bs on the migrated branch.

The acceptance question is: **does the conditions round-trip reproduce what the engine
actually sampled under?** The GRPO `ratio` cannot answer it — with
`old_logp_source: replay`, both π_old and the current-policy log-probs come from the
same replay path, so `ratio ≈ 1` by construction. So instead I ran the **same task on
both rollout engines** and compared reward curves; the only variable is the engine.

**PickScore A/B — 55 rollouts per arm** (Klein-4B, in-process PickScore, 1×H20 each),
identical dataset (same md5), seed, model bytes, venv, commit:

| | mean | slope / rollout | t (df=53) | verdict |
|---|---|---|---|---|
| SGLang | 0.7884 | −0.000023 | −0.41 | flat |
| Trainside (control) | 0.7886 | −0.000023 | −0.40 | flat |

Slopes match to six decimals, **r = 0.9983**, between-arm spread 17× tighter than each
arm's own rollout-to-rollout noise. Both curves are *flat* rather than growing, but that
is a property of the **reward** — PickScore scores text↔image alignment and never sees
the source image, so an edit task gives it little to reward.

**EditReward A/B — 60 rollouts per arm**, on a reward that *does* score edit quality
(`TIGER-Lab/EditReward-MiMo-VL-7B-SFT-2508`, Klein-4B, one dedicated scorer GPU + 7
training GPUs). wandb:
- baseline (trainside): https://wandb.ai/linyuwus/unirl-lin584-editreward-ab/runs/5ucncrko
- sglang: https://wandb.ai/linyuwus/unirl-lin584-editreward-ab/runs/yhgy1r1z

| | mean | slope / rollout | t (df=58) |
|---|---|---|---|
| SGLang | −0.292 | +0.00008 | +0.21 |
| Trainside (control) | −0.300 | +0.00060 | +1.41 |

**Cross-arm r = 0.93**, the arms move the **same direction on 52 of 59 rollout-to-rollout
steps (88%)**, between-arm spread 2.7× tighter than each arm's own noise. The small early
level offset (+0.025) decayed monotonically to +0.008 by rollout 60 — consistent with the
one documented confound (SGLang is forced to `init_same_noise: false`, which shifts the
level, not the slope).

**On growth:** *neither* arm grew significantly in 60 rollouts — a power limitation, not
an engine one. The proven recipe is configured for 10,000 rollouts and EditReward is
noisy (sd ≈ 0.05), so a small true slope cannot clear significance at n=60. A definitive
growth demonstration needs a longer run; it is out of scope for validating the engine
code, which the equivalence results already do.

**Bottom line:** two independent A/Bs — PickScore (r = 0.998) and EditReward (r = 0.93,
88% same-direction) — establish that this ti2i path produces the same
rollout→replay→update dynamics as the known-good trainside engine, and §A shows the
migrated code reproduces the conditioning tensors those A/Bs relied on.

### Recipe correctness (carried into both migrated recipes)

Standing up an end-to-end run earlier surfaced four defects in the *recipe*, not the
engine path. All are in both sglang recipes here:

| Fix | Why it mattered |
|---|---|
| `master_dtype: fp32` (with `param_dtype: bf16`) | An all-bf16 master rounds ~1e-3 grad-scale updates away (bf16 ulp ~4e-5 at \|w\|~0.01). The proven trainside recipe trains in fp32. |
| `activation_checkpointing: true` | Memory: each of the 8 actors colocates its own SGLang engine, and ti2i concatenates source-image tokens onto the noise sequence during replay. |
| `data_path` **requires** `DATA_PATH` | It previously defaulted to `datasets/image_edit/train.jsonl`, whose media uris are placeholders that do not resolve — the recipe failed at image load on rollout 1. Now it fails at compose with a clear `???`. |
| dropped `sampling.autocast_precision: bf16` | Redundant (already the default) and an unjustified divergence from the proven recipe. |

### Not validated

- **Numerical rollout↔replay log-prob parity.** With `old_logp_source: replay`, `ratio`
  shows replay is self-consistent, not that it matches what the engine sampled under. A
  real parity gate compares replay against the engine's emitted `trajectory_log_probs`.
  The §A round-trip proves the conditioning tensors are correct; it does not run replay.
- **LoRA weight sync under ti2i** (the round-trip ran `use_lora=False`).
- **base-9B** (the checkpoint pulled for all GPU work was 4B; both Klein variants inherit
  the identical TI2I condition path from `Flux2PipelineConfig`, which differs only in
  guidance embeddings and negative-CFG kwargs).
- **Full training-memory envelope of `batch_size: 64` (4B recipe)** — the value mirrors
  the trainside 4B recipe's larger batch; the round-trip used `forward_batch_size=4` and
  did not exercise the training path at that batch.

## Compatibility / Risk

- **Additive and opt-in.** The image branch is a no-op when `sample.has_image_input()`
  is false, so the existing t2i `flux2_klein` path is unchanged (regression-tested in
  §A/T3). No config migration; no checkpoint or data-format change.
- `patch_conditions` gains one captured field, gated on presence on the batch, so it
  stays `None` for every text-only rollout and for families that never set it.
- **`populate_conditions: true` is load-bearing for ti2i** (not merely a nicety as in
  t2i): replay has no source-image conditioning without it. Both recipes set it.
- The adapter keeps a local `first_per_group` / `_deexpand_prompts`, mirroring the
  Edit-Plus adapter (the established pattern for a per-sample payload that collapses
  alongside the prompts). No shared-module move in this PR.
- Throughput: SGLang rejects cross-request dynamic batching for image-bearing requests,
  so ti2i is somewhat slower per sample than t2i. `num_outputs_per_prompt` K-expansion
  still applies, so per-group efficiency is retained.

## Reviewer Notes

- **AI-assisted.** Adapter, patches, recipes, and the GPU round-trip harness were written
  with AI assistance; every line was reviewed, and the GPU validation in §A was run and
  inspected by the author.
- **Duplicate-work check:** no other open PR adds ti2i to `sglang_diffusion` for FLUX.2;
  the EditReward scorer batching fix is a separate PR (#269) and not part of this change.
- Inspect first: `build_prompts` / `build_condition` in
  `unirl/rollout/engine/sglang_diffusion/adapters/flux.py` and the
  `condition_image_latent_ids` capture in `_patches/patch_conditions.py`.

## Checklist

- [x] I reviewed the changed code and removed unrelated/generated artifacts.
- [x] I updated tests, docs, and configs where needed, or explained why not.

### CODE REVIEW

File: unirl/rollout/engine/sglang_diffusion/utils/prompts.py [hunk 1]
```diff
@@ -51,4 +51,22 @@ def deexpand_prompts_from_groups(
...
+    (``[A, A, B, B]``) or interleaved (``[A, B, A, B]``). Indexing ``items[::k]``
+    instead would misalign on the interleaved layout.
+    """
+    seen: set[str] = set()
+    out: List[Any] = []
+    for item, gid in zip(items, group_ids):
+        if gid not in seen:
+            seen.add(gid)
+            out.append(item)
+    return out
```
[the remark this answers is not in the document]
Comment (leviking98z-rgb) [reply unirl/rollout/engine/sglang_diffusion/utils/prompts.py]: Fixed in `dd30dd7`: `first_per_group` now validates equal lengths before iterating, and the new regression test covers both shorter-items and shorter-group_ids cases.

File: unirl/rollout/engine/sglang_diffusion/_patches/patch_conditions.py [hunk 2]
```diff
@@ -326,6 +341,26 @@ def _copy_conditions(src, output_batch) -> None:
...
+    if value is None:
+        return
+
+    import torch
+
+    if torch.is_tensor(value):
+        setattr(output_batch, dst, [value.detach().cpu()])
+    elif isinstance(value, (list, tuple)):
+        setattr(output_batch, dst, [t.detach().cpu() if torch.is_tensor(t) else t for t in value])
+
```
[the remark this answers is not in the document]
Comment (leviking98z-rgb) [reply unirl/rollout/engine/sglang_diffusion/_patches/patch_conditions.py]: Fixed in `dd30dd7`: `_copy_single_tensor_condition` now raises `TypeError` for unexpected source containers, with a regression covering the previous silent-drop case.

**Review by leviking98z-rgb — COMMENTED** [review]
Reviewed the full diff at head `12c8cf4` after bringing it up to date with `main`. The two outstanding fail-fast issues are fixed and covered: parallel payload/group lengths can no longer truncate silently, and unexpected condition containers can no longer be dropped silently. The request-side image/group alignment, condition-token/id capture, all-or-none response validation, and T2I no-op path are internally consistent; I found no remaining code-level blocker. Verified locally: 5 targeted pytest cases passed, targeted pre-commit passed (including recipe target resolution), and `git diff --check` is clean. Current lint/body/title/status checks are green; the PR body retains the author-recorded GPU rollout and A/B validation.

**Comment by celve** [comment]
@leviking98z-rgb A force push to resolve the sample migration. Could you please help me take a look again?
